File: modulo01-fundamentos-de-ia-e-llms-para-programadores/ecommerce-neural-recommender-java/src/main/java/com/recomendacao/ml/NeuralNetwork.java
package com.recomendacao.ml;

import java.util.Random;

/**
 * Rede Neural Feedforward implementada do zero em Java.
 *
 * Espelho da rede configurada em configureNeuralNetAndTrain() do original JS:
 *   - Camada entrada:   inputDim neurônios
 *   - Camada oculta 1: 128 neuronios, ReLU
 *   - Camada oculta 2:  64 neuronios, ReLU
 *   - Camada oculta 3:  32 neuronios, ReLU
 *   - Camada saida:      1 neuronio,  Sigmoid
 *
 * Algoritmo de treino: SGD (Stochastic Gradient Descent) com backpropagation.
 * Loss: Binary Cross-Entropy
 */
public class NeuralNetwork {

    // ====================================================================
    // Pesos e biases de cada camada
    // weights[l][i][j] = peso da camada l, neurônio i, entrada j
    // biases[l][i]     = bias da camada l, neurônio i
    // ====================================================================
    private final double[][][] weights;
    private final double[][] biases;
    private final String[] activations;
    private final double learningRate;

    // Cache para backpropagation
    private double[][] preActivations;  // valor ANTES da funcao de ativacao
    private double[][] postActivations; // valor DEPOIS da funcao de ativacao (saida)

    /**
     * @param layerSizes    tamanho de cada camada, ex: [20, 128, 64, 32, 1]
     * @param learningRate  taxa de aprendizado (ex: 0.01)
     */
    public NeuralNetwork(int[] layerSizes, double learningRate) {
        this.learningRate = learningRate;
        int numLayers = layerSizes.length - 1;

        weights     = new double[numLayers][][];
        biases      = new double[numLayers][];
        activations = new String[numLayers];
        preActivations  = new double[numLayers][];
        postActivations = new double[numLayers + 1][];

        Random rand = new Random(42); // seed fixo para reproducibilidade

        for (int l = 0; l < numLayers; l++) {
            int in  = layerSizes[l];
            int out = layerSizes[l + 1];

            weights[l] = new double[out][in];
            biases[l]  = new double[out];

            // Inicializacao He: escala = sqrt(2 / entradas)
            // Funciona bem com ReLU, evita vanishing/exploding gradients
            double scale = Math.sqrt(2.0 / in);
            for (int i = 0; i < out; i++) {
                for (int j = 0; j < in; j++) {
                    weights[l][i][j] = rand.nextGaussian() * scale;
                }
            }

            // Ultima camada usa sigmoid, demais usam relu
            activations[l] = (l == numLayers - 1) ? "sigmoid" : "relu";
        }
    }

    // ====================================================================
    // Funcoes de ativacao
    // ====================================================================

    private double relu(double x) {
        return Math.max(0.0, x);
    }

    private double sigmoid(double x) {
        return 1.0 / (1.0 + Math.exp(-x));
    }

    /** Derivada de ReLU em relacao ao pre-ativacao */
    private double reluDeriv(double preAct) {
        return preAct > 0 ? 1.0 : 0.0;
    }

    /** Derivada de Sigmoid em relacao ao pre-ativacao */
    private double sigmoidDeriv(double preAct) {
        double s = sigmoid(preAct);
        return s * (1.0 - s);
    }

    // ====================================================================
    // Forward pass: calcula a predicao dado um vetor de entrada
    // ====================================================================

    /**
     * Propaga a entrada pela rede e retorna a predicao (0-1).
     * Salva os valores intermediarios para o backpropagation.
     */
    public double forward(double[] input) {
        postActivations[0] = input;
        double[] current = input;

        for (int l = 0; l < weights.length; l++) {
            int outSize = weights[l].length;
            double[] pre  = new double[outSize];
            double[] post = new double[outSize];

            for (int i = 0; i < outSize; i++) {
                pre[i] = biases[l][i];
                for (int j = 0; j < current.length; j++) {
                    pre[i] += weights[l][i][j] * current[j];
                }
                post[i] = "relu".equals(activations[l]) ? relu(pre[i]) : sigmoid(pre[i]);
            }

            preActivations[l]      = pre;
            postActivations[l + 1] = post;
            current = post;
        }

        return current[0]; // saida binaria unica
    }

    // ====================================================================
    // Backward pass: calcula gradientes e atualiza os pesos
    // ====================================================================

    /**
     * Backpropagation com SGD.
     * Chame sempre APOS um forward() com o mesmo exemplo.
     *
     * @param label valor real (0 ou 1)
     */
    public void backward(double label) {
        int numLayers = weights.length;
        double[][] deltas = new double[numLayers][];

        // --- Camada de saida ---
        // delta = (predicao - label) * sigmoid'(preAct)
        // derivada da binary cross-entropy + sigmoid juntas simplifica para: (pred - label)
        double pred   = postActivations[numLayers][0];
        double preOut = preActivations[numLayers - 1][0];
        deltas[numLayers - 1] = new double[]{ (pred - label) * sigmoidDeriv(preOut) };

        // --- Camadas ocultas (de tras para frente) ---
        for (int l = numLayers - 2; l >= 0; l--) {
            int size = weights[l].length;
            deltas[l] = new double[size];

            for (int i = 0; i < size; i++) {
                double sum = 0;
                for (int k = 0; k < deltas[l + 1].length; k++) {
                    sum += weights[l + 1][k][i] * deltas[l + 1][k];
                }
                deltas[l][i] = sum * reluDeriv(preActivations[l][i]);
            }
        }

        // --- Atualizar pesos e biases ---
        for (int l = 0; l < numLayers; l++) {
            double[] prevOutput = postActivations[l];
            for (int i = 0; i < weights[l].length; i++) {
                for (int j = 0; j < weights[l][i].length; j++) {
                    weights[l][i][j] -= learningRate * deltas[l][i] * prevOutput[j];
                }
                biases[l][i] -= learningRate * deltas[l][i];
            }
        }
    }

    // ====================================================================
    // Treino completo (equivalente ao model.fit() do TensorFlow.js)
    // ====================================================================

    /**
     * Treina a rede neural.
     *
     * @param inputs  matriz de entrada [numExemplos][numFeatures]
     * @param labels  rotulos binarios [numExemplos]
     * @param epochs  numero de epocas
     */
    public void train(double[][] inputs, double[] labels, int epochs) {
        System.out.println("\n=== Iniciando Treinamento da Rede Neural ===");
        System.out.printf("Exemplos: %d | Epocas: %d | Learning Rate: %.3f%n%n",
                inputs.length, epochs, learningRate);

        Random rand = new Random();
        int n = inputs.length;

        for (int epoch = 0; epoch < epochs; epoch++) {
            double totalLoss = 0;
            int correct = 0;

            // Embaralhar indices a cada epoca (equivalente ao shuffle: true do TF.js)
            int[] indices = createShuffledIndices(n, rand);

            for (int idx : indices) {
                double pred  = forward(inputs[idx]);
                double label = labels[idx];

                backward(label);

                // Binary cross-entropy: -[y*log(p) + (1-y)*log(1-p)]
                double loss = -(label * Math.log(pred + 1e-7)
                             + (1 - label) * Math.log(1 - pred + 1e-7));
                totalLoss += loss;

                if ((pred >= 0.5 && label == 1.0) || (pred < 0.5 && label == 0.0)) {
                    correct++;
                }
            }

            if ((epoch + 1) % 10 == 0) {
                double avgLoss = totalLoss / n;
                double accuracy = (double) correct / n * 100;
                System.out.printf("Epoca %3d | Loss: %.4f | Acuracia: %.1f%%%n",
                        epoch + 1, avgLoss, accuracy);
            }
        }

        System.out.println("\n=== Treinamento Concluido ===\n");
    }

    private int[] createShuffledIndices(int n, Random rand) {
        int[] indices = new int[n];
        for (int i = 0; i < n; i++) indices[i] = i;
        for (int i = n - 1; i > 0; i--) {
            int j = rand.nextInt(i + 1);
            int tmp = indices[i]; indices[i] = indices[j]; indices[j] = tmp;
        }
        return indices;
    }
}

File: modulo01-fundamentos-de-ia-e-llms-para-programadores/ecommerce-neural-recommender-go/ml/neural_network.go
// Package ml implementa a rede neural feedforward do zero em Go — porte de
// NeuralNetwork.java, ModelContext.java e RecommendationEngine.java.
package ml

import (
	"fmt"
	"math"
	"math/rand"
)

// NeuralNetwork é uma rede feedforward treinada via SGD + backpropagation,
// com loss de Binary Cross-Entropy — mesma arquitetura da versão Java:
// camadas ocultas com ReLU, camada de saída com Sigmoid.
type NeuralNetwork struct {
	// weights[l][i][j] = peso da camada l, neurônio i, entrada j.
	weights [][][]float64
	biases  [][]float64
	// activations[l] é "relu" para camadas ocultas e "sigmoid" para a última.
	activations  []string
	learningRate float64

	// Cache para backpropagation.
	preActivations  [][]float64 // valor ANTES da função de ativação
	postActivations [][]float64 // valor DEPOIS da função de ativação (saída)

	rng *rand.Rand
}

// NewNeuralNetwork cria a rede com o formato de camadas informado (ex.:
// [20, 128, 64, 32, 1]) e a taxa de aprendizado dada. Usa seed fixo (42) na
// inicialização dos pesos, para reprodutibilidade — igual à versão Java.
func NewNeuralNetwork(layerSizes []int, learningRate float64) *NeuralNetwork {
	numLayers := len(layerSizes) - 1

	nn := &NeuralNetwork{
		weights:         make([][][]float64, numLayers),
		biases:          make([][]float64, numLayers),
		activations:     make([]string, numLayers),
		preActivations:  make([][]float64, numLayers),
		postActivations: make([][]float64, numLayers+1),
		learningRate:    learningRate,
		rng:             rand.New(rand.NewSource(42)),
	}

	for l := 0; l < numLayers; l++ {
		in := layerSizes[l]
		out := layerSizes[l+1]

		nn.weights[l] = make([][]float64, out)
		nn.biases[l] = make([]float64, out)

		// Inicialização He: escala = sqrt(2 / entradas). Funciona bem com
		// ReLU, evita vanishing/exploding gradients.
		scale := math.Sqrt(2.0 / float64(in))
		for i := 0; i < out; i++ {
			nn.weights[l][i] = make([]float64, in)
			for j := 0; j < in; j++ {
				nn.weights[l][i][j] = nn.rng.NormFloat64() * scale
			}
		}

		if l == numLayers-1 {
			nn.activations[l] = "sigmoid"
		} else {
			nn.activations[l] = "relu"
		}
	}

	return nn
}

func relu(x float64) float64 {
	return math.Max(0, x)
}

func sigmoid(x float64) float64 {
	return 1.0 / (1.0 + math.Exp(-x))
}

func reluDeriv(preAct float64) float64 {
	if preAct > 0 {
		return 1.0
	}
	return 0.0
}

func sigmoidDeriv(preAct float64) float64 {
	s := sigmoid(preAct)
	return s * (1.0 - s)
}

// Forward propaga a entrada pela rede e retorna a predição (0-1), salvando
// os valores intermediários necessários para o Backward subsequente.
func (nn *NeuralNetwork) Forward(input []float64) float64 {
	nn.postActivations[0] = input
	current := input

	for l := 0; l < len(nn.weights); l++ {
		outSize := len(nn.weights[l])
		pre := make([]float64, outSize)
		post := make([]float64, outSize)

		for i := 0; i < outSize; i++ {
			pre[i] = nn.biases[l][i]
			for j := 0; j < len(current); j++ {
				pre[i] += nn.weights[l][i][j] * current[j]
			}
			if nn.activations[l] == "relu" {
				post[i] = relu(pre[i])
			} else {
				post[i] = sigmoid(pre[i])
			}
		}

		nn.preActivations[l] = pre
		nn.postActivations[l+1] = post
		current = post
	}

	return current[0] // saída binária única
}

// Backward executa backpropagation com SGD. Deve ser chamado sempre APÓS um
// Forward() com o mesmo exemplo.
func (nn *NeuralNetwork) Backward(label float64) {
	numLayers := len(nn.weights)
	deltas := make([][]float64, numLayers)

	// Camada de saída: a derivada da binary cross-entropy + sigmoid juntas
	// simplifica para (pred - label).
	pred := nn.postActivations[numLayers][0]
	preOut := nn.preActivations[numLayers-1][0]
	deltas[numLayers-1] = []float64{(pred - label) * sigmoidDeriv(preOut)}

	// Camadas ocultas, de trás para frente.
	for l := numLayers - 2; l >= 0; l-- {
		size := len(nn.weights[l])
		deltas[l] = make([]float64, size)

		for i := 0; i < size; i++ {
			sum := 0.0
			for k := 0; k < len(deltas[l+1]); k++ {
				sum += nn.weights[l+1][k][i] * deltas[l+1][k]
			}
			deltas[l][i] = sum * reluDeriv(nn.preActivations[l][i])
		}
	}

	// Atualizar pesos e biases.
	for l := 0; l < numLayers; l++ {
		prevOutput := nn.postActivations[l]
		for i := range nn.weights[l] {
			for j := range nn.weights[l][i] {
				nn.weights[l][i][j] -= nn.learningRate * deltas[l][i] * prevOutput[j]
			}
			nn.biases[l][i] -= nn.learningRate * deltas[l][i]
		}
	}
}

// Train treina a rede por `epochs` épocas, embaralhando os índices a cada
// época (equivalente ao shuffle: true do TF.js) e imprimindo loss/acurácia a
// cada 10 épocas — mesmo comportamento observável da versão Java.
func (nn *NeuralNetwork) Train(inputs [][]float64, labels []float64, epochs int) {
	fmt.Println()
	fmt.Println("=== Iniciando Treinamento da Rede Neural ===")
	fmt.Printf("Exemplos: %d | Epocas: %d | Learning Rate: %.3f\n\n", len(inputs), epochs, nn.learningRate)

	n := len(inputs)
	shuffleRng := rand.New(rand.NewSource(rand.Int63()))

	for epoch := 0; epoch < epochs; epoch++ {
		totalLoss := 0.0
		correct := 0

		indices := shuffledIndices(n, shuffleRng)

		for _, idx := range indices {
			pred := nn.Forward(inputs[idx])
			label := labels[idx]

			nn.Backward(label)

			loss := -(label*math.Log(pred+1e-7) + (1-label)*math.Log(1-pred+1e-7))
			totalLoss += loss

			if (pred >= 0.5 && label == 1.0) || (pred < 0.5 && label == 0.0) {
				correct++
			}
		}

		if (epoch+1)%10 == 0 {
			avgLoss := totalLoss / float64(n)
			accuracy := float64(correct) / float64(n) * 100
			fmt.Printf("Epoca %3d | Loss: %.4f | Acuracia: %.1f%%\n", epoch+1, avgLoss, accuracy)
		}
	}

	fmt.Println()
	fmt.Println("=== Treinamento Concluido ===")
	fmt.Println()
}

func shuffledIndices(n int, rng *rand.Rand) []int {
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	rng.Shuffle(n, func(i, j int) {
		indices[i], indices[j] = indices[j], indices[i]
	})
	return indices
}
